### PR TITLE
kafka(ticdc): kafka sink use txn appender

### LINKED ISSUE #9472: kafka(ticdc): kafka sink use txn appender
kafka(ticdc): kafka sink use txn appender

### PR DESCRIPTION
**Pull request by sdojjy**
### What problem does this PR solve?

Issue Number: close #9472

### What is changed and how it works?
kafka sink use txn appender to send the changefeed rows to table sink,  then table sink can rebuild and re-order the whole txn.

#### Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

### Release note 

```release-note
 `None`.
```

### CODE REVIEW

File: cdc/sink/dmlsink/mq/mq_dml_sink.go [hunk 1]
```diff
@@ -119,35 +120,49 @@ func newDMLSink(
...
-		partition := s.alive.eventRouter.GetPartitionForRowChange(row.Event, partitionNum)
-		// This never be blocked because this is an unbounded channel.
-		s.alive.worker.msgChan.In() <- mqEvent{
-			key: TopicPartitionKey{
-				Topic: topic, Partition: partition,
-			},
-			rowEvent: row,
+		callback := mergedCallback(txn.Callback, uint64(len(txn.Event.Rows)))
+		for _, row := range txn.Event.Rows {
+			topic := s.alive.eventRouter.GetTopicForRowChange(row)
```
Comment (3AceShowHand) [inline cdc/sink/dmlsink/mq/mq_dml_sink.go:147]: It looks that `for one transactions, all rows belongs to the same table, have the same commitTs, so should be delivered to the same topic and partition.`

File: cdc/sink/dmlsink/mq/mq_dml_sink.go [hunk 2]
```diff
@@ -119,35 +120,49 @@ func newDMLSink(
...
-			return errors.Trace(err)
-		}
-		partition := s.alive.eventRouter.GetPartitionForRowChange(row.Event, partitionNum)
-		// This never be blocked because this is an unbounded channel.
-		s.alive.worker.msgChan.In() <- mqEvent{
-			key: TopicPartitionKey{
-				Topic: topic, Partition: partition,
-			},
-			rowEvent: row,
+		callback := mergedCallback(txn.Callback, uint64(len(txn.Event.Rows)))
```
Comment (3AceShowHand) [inline cdc/sink/dmlsink/mq/mq_dml_sink.go:145]: It looks this `callback` is shared among all row events.

**Review by 3AceShowHand — APPROVED** [review]

File: cdc/sink/dmlsink/mq/mq_dml_sink.go (hunk 2 above)
Comment (sdojjy) [reply cdc/sink/dmlsink/mq/mq_dml_sink.go:145]: yes,  for txn appender, there is only one callback,  we must split  and merge that callback in kafka sink

File: cdc/sink/dmlsink/mq/mq_dml_sink.go (hunk 1 above)
Comment (sdojjy) [reply cdc/sink/dmlsink/mq/mq_dml_sink.go:147]: the topic and partition are based on row value and dispatch rule?

File: cdc/sink/dmlsink/mq/mq_dml_sink.go [hunk 3]
```diff
@@ -119,35 +120,49 @@ func newDMLSink(
 
 // WriteEvents writes events to the sink.
 // This is an asynchronously and thread-safe method.
-func (s *dmlSink) WriteEvents(rows ...*dmlsink.RowChangeCallbackableEvent) error {
+func (s *dmlSink) WriteEvents(txns ...*dmlsink.CallbackableEvent[*model.SingleTableTxn]) error {
```
Comment (nongfushanquan) [inline cdc/sink/dmlsink/mq/mq_dml_sink.go:123]: Should it handle the big transaction problem ?

File: cdc/sink/dmlsink/mq/mq_dml_sink.go (hunk 2 above)
Comment (asddongmen) [reply cdc/sink/dmlsink/mq/mq_dml_sink.go:145]: so, when we need to call the only one callback?

File: cdc/sink/dmlsink/mq/mq_dml_sink.go (hunk 3 above)
Comment (sdojjy) [reply cdc/sink/dmlsink/mq/mq_dml_sink.go:123]: big transaction is already split, like MySQL sink

File: cdc/sink/dmlsink/mq/mq_dml_sink.go (hunk 2 above)
Comment (sdojjy) [reply cdc/sink/dmlsink/mq/mq_dml_sink.go:145]: if all events are callbacked, then we call the original callback

**Review by nongfushanquan — APPROVED** [review]

**Pull request merged**